File: backend/src/main/java/com/ares/backend/entity/Receta.java
package com.ares.backend.entity;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.Filter;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

/**
 * Entidad que representa una receta de cocina.
 * Contiene los pasos a seguir y los ingredientes necesarios.
 * Solo puede ser creada y modificada por jefes de cocina.
 *
 * @author Ares
 * @version 1.0
 */
@Entity
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@ToString
@Table(name = "recetas")
@Filter(name = "negocioFilter", condition = "negocio_id = :negocioId")
public class Receta {

    /**
     * Identificador único de la receta.
     */
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    /**
     * Negocio (tenant) al que pertenece esta receta. Las tablas hijas
     * (pasos_receta, receta_ingredientes, recetas_favoritas) se scopean
     * transitivamente a través de esta receta, sin columna propia.
     * Endurecido a nullable=false en Fase 9/PR6, una vez verificado que
     * RecetaService setea siempre negocio al crear — consistente con la
     * constraint NOT NULL de la migración V2 (BD).
     */
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "negocio_id", nullable = false)
    private Negocio negocio;

    /**
     * Nombre de la receta (ej: Pasta Carbonara, Ensalada César).
     */
    @Column(nullable = false, length = 100)
    private String nombre;

    /**
     * Descripción general de la receta.
     */
    @Column(length = 500)
    private String descripcion;

    /**
     * Fecha y hora de creación de la receta.
     */
    @Column(nullable = false)
    private LocalDateTime fechaCreacion;

    /**
     * Usuario (jefe de cocina) que creó la receta.
     */
    @ManyToOne
    @JoinColumn(name = "creada_por_id", nullable = false)
    private Usuario creadaPor;

    /**
     * Lista de pasos ordenados para elaborar la receta.
     */
    @OneToMany(mappedBy = "receta", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<PasoReceta> pasos = new ArrayList<>();

    /**
     * Lista de ingredientes necesarios con sus cantidades.
     */
    @OneToMany(mappedBy = "receta", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<RecetaIngrediente> ingredientes = new ArrayList<>();

    /**
     * Constructor con parámetros para crear una receta.
     *
     * @param nombre Nombre de la receta
     * @param descripcion Descripción de la receta
     * @param creadaPor Usuario jefe de cocina que crea la receta
     */
    public Receta(String nombre, String descripcion, Usuario creadaPor) {
        this.nombre = nombre;
        this.descripcion = descripcion;
        this.creadaPor = creadaPor;
        this.fechaCreacion = LocalDateTime.now();
    }
}

File: frontend/flutter_app/lib/models/receta.dart
import 'paso_receta.dart';
import 'receta_ingrediente.dart';

/// Receta completa con sus pasos e ingredientes embebidos.
///
/// [creadaPorId] referencia al [Usuario] jefe de cocina que la creó.
/// [puedElaborarse] es calculado por el backend al verificar stock;
/// en el modelo Flutter se almacena como campo opcional para cachear
/// el resultado de la última verificación.
class Receta {
  final int                   id;
  final String                nombre;
  final String                descripcion;
  final List<PasoReceta>      pasos;
  final List<RecetaIngrediente> ingredientes;
  final int                   creadaPorId;
  final String                fechaCreacion; // ISO-8601
  final bool?                 puedeElaborarse; // null = no verificado aún

  const Receta({
    required this.id,
    required this.nombre,
    required this.descripcion,
    required this.pasos,
    required this.ingredientes,
    required this.creadaPorId,
    required this.fechaCreacion,
    this.puedeElaborarse,
  });

  factory Receta.fromJson(Map<String, dynamic> json) => Receta(
        id:              json['id']          as int,
        nombre:          json['nombre']      ?? '',
        descripcion:     json['descripcion'] ?? '',
        pasos:           json['pasos'] != null 
            ? (json['pasos'] as List<dynamic>)
                .map((e) => PasoReceta.fromJson(e as Map<String, dynamic>))
                .toList()
            : [],
        ingredientes:    json['ingredientes'] != null
            ? (json['ingredientes'] as List<dynamic>)
                .map((e) => RecetaIngrediente.fromJson(e as Map<String, dynamic>))
                .toList()
            : [],
        creadaPorId:     json['creadaPor'] != null ? (json['creadaPor']['id'] as int) : 0,
        fechaCreacion:   json['fechaCreacion']   ?? '',
        puedeElaborarse: json['disponible'] as bool?,
      );

  Map<String, dynamic> toJson() => {
        'id':              id,
        'nombre':          nombre,
        'descripcion':     descripcion,
        'pasos':           pasos.map((p) => p.toJson()).toList(),
        'ingredientes':    ingredientes.map((i) => i.toJson()).toList(),
        'creadaPorId':     creadaPorId,
        'fechaCreacion':   fechaCreacion,
        'disponible':      puedeElaborarse,
      };

  Receta copyWith({
    int?                    id,
    String?                 nombre,
    String?                 descripcion,
    List<PasoReceta>?       pasos,
    List<RecetaIngrediente>? ingredientes,
    int?                    creadaPorId,
    String?                 fechaCreacion,
    bool?                   puedeElaborarse,
  }) =>
      Receta(
        id:              id              ?? this.id,
        nombre:          nombre          ?? this.nombre,
        descripcion:     descripcion     ?? this.descripcion,
        pasos:           pasos           ?? this.pasos,
        ingredientes:    ingredientes    ?? this.ingredientes,
        creadaPorId:     creadaPorId     ?? this.creadaPorId,
        fechaCreacion:   fechaCreacion   ?? this.fechaCreacion,
        puedeElaborarse: puedeElaborarse ?? this.puedeElaborarse,
      );
}